### PR TITLE
Refactor the external API to use Pydantic models

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description

This PR is a follow up to #102 and implements the API changes described in #49. It completes the effort to make fragmenters stateless, such that molecules are now passed directly to ``fragment``:

```
fragmenter = WBOFragmenter()

result = fragmenter.fragment(
    Molecule.from_smiles(
        "CC(C)(C)c1sc(c2ccnc(N)n2)c(n1)c3cccc(N[S](=O)(=O)c4c(F)cccc4F)c3F"
    )
)
```

and the generated fragments are now returned as a new object that also contains provenance information. 

This PR further solidifies the options that can be passed to a fragmenter by now forbidding passing `kwargs`. The result is a much more well defined and easy to document API. Fragmentation settings may either be defined when defining the fragmentation engine:

```
fragmenter = WBOFragmenter(threshold=0.05)
```
or after it's been initialized:
```
fragmenter.threshold = 0.07
```

Because engines are now pydantic based they can be directly (de)serialised:

```
>>> fragmenter = WBOFragmenter()
>>> fragmenter.json()

{
  "functional_groups": {
    "hydrazine": "[NX3:1][NX3:2]",
    ...
    "tri_halide": "[#6:1]([F,Cl,I,Br:2])([F,Cl,I,Br:3])([F,Cl,I,Br:4])"
  },
  "scheme": "WBO",
  "wbo_options": {
    "method": "am1-wiberg-elf10",
    "max_conformers": 800,
    "rms_threshold": 1.0
  },
  "threshold": 0.03,
  "heuristic": "path_length",
  "keep_non_rotor_ring_substituents": false
}
```

allowing them to be easily included in other workflow schemas (e.g. the bespoke fit optimization schema).

The returned result object is also pydantic based and hence can also be easily serialized:

```
>>> result.json()

{
  "parent_smiles": "[H:31][c:1]1[c:3]([c:9]...",
  "fragments": [
    {
      "smiles": "[H:34][c:4]1[c:12]([c:5]...",
      "bond_indices": [8, 24]
    },
    ...
  ],
  "provenance": {
    "creator": "fragmenter",
    "version": "0.0.7+4.g8735891.dirty",
    "options": {
      "functional_groups": {
        "hydrazine": "[NX3:1][NX3:2]",
        ...
        "tri_halide": "[#6:1]([F,Cl,I,Br:2])([F,Cl,I,Br:3])([F,Cl,I,Br:4])"
      },
      "scheme": "WBO",
      "wbo_options": {
        "method": "am1-wiberg-elf10",
        "max_conformers": 800,
        "rms_threshold": 1.0
      },
      "threshold": 0.03,
      "heuristic": "path_length",
      "keep_non_rotor_ring_substituents": false
    }
  }
}
```

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by SimonBoothroyd** [comment]
@j-wags @jthorton @lilyminium @jchodera I'd like to get this merged in ASAP so we can move onto the full regression tests and get a new C-F release made. If anyone has any objects could you please let me know by the end of today. I'll plan to merge this first thing tomorrow otherwise!

File: docs/api.rst [hunk 1]
```diff
@@ -4,16 +4,41 @@ API
 Below is an outline of the API for the main functions of ``fragmenter`` See :doc:`examples <examples>` for details on
 how to use these functions.
 
-.. warning:: ``fragmenter`` The ``fragmenter`` package is still pre-alpha so the API is still in flux.is still pre-alpha so the API is still in flux.
+.. warning:: ``fragmenter`` The ``fragmenter`` package is still pre-alpha so the API is still in flux.is still
```
Comment (lilyminium) [inline docs/api.rst:7]:
```suggestion
.. warning:: ``fragmenter`` The ``fragmenter`` package is still pre-alpha so the API is still
```

File: fragmenter/fragment.py [hunk 2]
```diff
@@ -29,37 +31,75 @@
...
+
+    bond_indices: Tuple[int, int] = Field(
+        ...,
+        description="The map indices of the atoms involved in the bond that the "
+        "fragment was built around.",
+    )
+
+    @property
+    def molecule(self) -> Molecule:
+        """The fragmented represented as an OpenFF molecule object."""
```
Comment (lilyminium) [inline fragmenter/fragment.py:54]:
```suggestion
        """The fragment represented as an OpenFF molecule object."""
```

File: fragmenter/fragment.py [hunk 3]
```diff
@@ -29,37 +31,75 @@
...
+
+    @property
+    def parent_molecule(self) -> Molecule:
+        """The parent molecule represented as an OpenFF molecule object."""
+        return Molecule.from_smiles(self.parent_smiles)
+
+    @property
+    def fragment_molecules(self) -> Dict[BondTuple, Molecule]:
+        """A dictionary of the fragment molecules represented as OpenFF molecule
+        objects indexed by the map indices of the bond that each fragment was built
```
Comment (lilyminium) [inline fragmenter/fragment.py:83]:
```suggestion
        objects, indexed by the map indices of the bond that each fragment was built
```

File: fragmenter/fragment.py [hunk 4]
```diff
@@ -29,37 +31,75 @@
...
 
-        self.functional_groups = functional_groups
+class Fragmenter(BaseModel, abc.ABC):
+    """The base class that all fragmentation engines should inherit from."""
 
-        self.fragments = {}
+    functional_groups: Dict[str, str] = Field(
+        default_factory=get_fgroup_smarts,
+        description="A dictionary of SMARTS of functional groups that should not be "
+        "fragmented indexed by an informative name, e.g. 'alcohol': '[#6]-[#8X2H1]'.",
```
Comment (lilyminium) [inline fragmenter/fragment.py:101]:
```suggestion
        "fragmented, indexed by an informative name, e.g. 'alcohol': '[#6]-[#8X2H1]'.",
```

File: fragmenter/fragment.py [hunk 5]
```diff
@@ -766,122 +806,128 @@ def _prepare_molecule(
         return molecule, stereo, found_functional_groups, found_ring_systems
 
     @abc.abstractmethod
-    def fragment(self, **kwargs):
-        """Fragment molecules."""
-        raise NotImplementedError()
+    def fragment(self, molecule: Molecule) -> FragmentationResult:
+        """Fragments a molecule according to this classes settings.
```
Comment (lilyminium) [inline fragmenter/fragment.py:810]:
```suggestion
        """Fragments a molecule according to this class' settings.
```

File: fragmenter/fragment.py [hunk 6]
```diff
@@ -766,122 +806,128 @@ def _prepare_molecule(
...
-        # import socket
-        # import uuid
-        #
-        # import fragmenter
-        #
-        # fragmenter_version = fragmenter.__version__
+
+        raise NotImplementedError()
+
+    def _default_provenance(self) -> Dict[str, Any]:
```
Comment (lilyminium) [inline fragmenter/fragment.py:825]: Will there be other methods of getting provenance, therefore requiring this one to specify default?

**Review by lilyminium — COMMENTED** [review]
Sorry @SimonBoothroyd, didn't have much time to do anything other than skim.

File: docs/api.rst (hunk 1 above)
Comment (SimonBoothroyd) [reply docs/api.rst:7]:
```suggestion
.. warning:: ``fragmenter`` The ``fragmenter`` package is still pre-alpha so the API is still in flux.
```

File: fragmenter/fragment.py (hunk 6 above)
Comment (SimonBoothroyd) [reply fragmenter/fragment.py:825]: I think the subclasses may in future need to store extra provenance (e.g. the WBO values for the rotors for the WBOFragmenter), so I would consider this the 'default that must be included'.

**Comment by SimonBoothroyd** [comment]
Thanks @lilyminium !

**Pull request merged**